### ISSUE TITLE
Support for unlock when MI Band (or other known) BLE device is nearby

### ISSUE THREAD

**Issue by cpfeiffer** [Freeyourgadget/Gadgetbridge#34]
With Android 5.0 it is possible to bypass the screen lock. We should provide such an option when e.g. the MI Band is nearby (need to check RSSI value).

**Comment by cpfeiffer** [comment]
It looks like this won't be easily doable, because this feature needs support from the ROM distributor. So called "Trust Agents" can only be provided by the ROM distributor, c.f. http://nelenkov.blogspot.de/2014/12/dissecting-lollipops-smart-lock.html

So with plain 5.0 we cannot provide our own Trust Agent. Maybe the default (Google) "Smart Lock" Trust Agent can already do what we want, but I don't have it on my phone.

**Comment by xphnx** [comment]
Maybe this could help https://github.com/npike/AndroidWearUnlock

**Comment by cpfeiffer** [comment]
Thanks for the pointer, that is indeed an interesting approach, but none that I'm feeling very comfortable with. It basically sets an empty password to keep the device unlocked while the wear-device is present.
Later it sets the password back in order to properly lock the device.

I hope we're going to see better support for trust agents, at least in custom roms.

**Comment by danielegobbetti** [comment]
I also don't like the approach of the empty password. Perhaps it could be explored if "Device Administrator" apps are allowed to set Trust agents as well as the distributor. Or we could get in touch with the omnirom / other rom / ... developers and see if there is interest for supporting GadgetBridge. :)

**Comment by cpfeiffer** [comment]
Anyone with contacts to Cyanogendmod, Omnirom etc.? :-)

**Comment by arnottcr** [comment]
I used to use google's trusted device agent, but have since moved to Cyanogenmod's System profiles (they allow for much more control and customization). This works out of the box, as the mi band fully pairs with the the phone gadget bridge is running on.

I can provide config help, but it is pretty straightforward. Honestly the mi band is one of the few that just works as a bluetooth tether.

**Comment by cpfeiffer** [comment]
AFAIR, the profiles only allow automatic unlock when the lock mode is a pattern. As soon as you have configured a pin, or if you have encrypted your device, automatic unlock is forbidden.

Can you confirm this?

**Comment by arnottcr** [comment]
So, profiles will allow you to override pins, patterns and passwords, but there is an interesting _issue_ related to encryption. I have created [a thread](http://forum.cyanogenmod.org/topic/118825-cm13-system-profiles-encryption-bug/) on the CM forums, but it has gotten no traction. There is more information there, but the issue has a workaround as long as you are fine with "Insecure" mode, and do not require "Disable" mode.

**Comment by cpfeiffer** [comment]
@arnottcr 
No, only patterns may be overridden by profiles, AFAICS, at least on my device. It says "This option is locked by the system administrator" (translated to English by me).

And this is apparently not a bug, but a feature. If the user wants more security, e.g. by encrypting the device, using a pin or password, then he should not use profiles to effectively disable all these security measures.

**Comment by xylo** [comment]
@arnottcr 
Thanks for the tip with the profiles!  Works like a charm.
